src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -1322,33 +1344,53 @@ protected void SetOutputPathAndRelatedData(string objOutputPath)
 
             // refresh final output path
             string newBinOutputPath;
-            if (TryGetOutputPathFromBuildManager(out newBinOutputPath) && newBinOutputPath != null)
+            if (hasSameBinAndObjOutputPaths)
             {
-                if (!string.Equals(_binOutputPathOpt, newBinOutputPath, StringComparison.OrdinalIgnoreCase))
-                {
-                    string oldBinOutputPath = _binOutputPathOpt;
+                newBinOutputPath = objOutputPath;
+            }
+            else if (!TryGetOutputPathFromBuildManager(out newBinOutputPath))
+            {
+                newBinOutputPath = null;
+            }
 
-                    // set obj output path if changed
-                    _binOutputPathOpt = newBinOutputPath;
+            if (newBinOutputPath != null && !string.Equals(_binOutputPathOpt, newBinOutputPath, StringComparison.OrdinalIgnoreCase))
+            {
+                string oldBinOutputPath = _binOutputPathOpt;
 
+                // set obj output path if changed
+                _binOutputPathOpt = newBinOutputPath;
+
+                // If the project has been hooked up with the project tracker, then update the bin path with the tracker.
+                if (this.ProjectTracker.GetProject(Id) != null)
+                {
                     this.ProjectTracker.UpdateProjectBinPath(this, oldBinOutputPath, _binOutputPathOpt);
                 }
             }
         }
 
-        private void UpdateProjectDisplayNameAndFilePath()
+        private static string GetProjectDisplayName(IVsHierarchy hierarchy)
         {
-            bool updateMade = false;
             string newDisplayName;
-            if (TryGetProjectDisplayName(_hierarchy, out newDisplayName) && this.DisplayName != newDisplayName)
+            return TryGetProjectDisplayName(hierarchy, out newDisplayName) ? newDisplayName : null;
+        }
+
+        private static string GetProjectFilePath(IVsHierarchy hierarchy)
+        {
+            string filePath;
+            return ErrorHandler.Succeeded(((IVsProject3)hierarchy).GetMkDocument((uint)VSConstants.VSITEMID.Root, out filePath)) ? filePath : null;
+        }
+
+        protected void UpdateProjectDisplayNameAndFilePath(string newDisplayName, string newPath)
+        {
+            bool updateMade = false;
+            
+            if (newDisplayName != null && this.DisplayName != newDisplayName)
             {
                 this.DisplayName = newDisplayName;
                 updateMade = true;
             }
-
-            string newPath;
-            if (ErrorHandler.Succeeded(((IVsProject3)_hierarchy).GetMkDocument((uint)VSConstants.VSITEMID.Root, out newPath)) &&
-                File.Exists(newPath) && _filePathOpt != newPath)
+            
+            if (newPath != null && File.Exists(newPath) && _filePathOpt != newPath)
             {
                 Debug.Assert(PathUtilities.IsAbsolute(newPath));
```

Comment:
I didn't want to change the existing behavior of AbstractProject, we can probably add this validation in our sub-type of AbstractProject and handle it early.